proposals/0351-regex-builder.md

```diff
@@ -428,16 +441,204 @@ extension RegexComponentBuilder {
+Capture {
+  ChoiceOf {
+    "CREDIT"
+    "DEBIT"
+  }
+} transform: {
+  "Transaction Kind: \($0)"
+} // `.RegexOutput == (Substring, String)`
+```
+
+The transform closure can throw. When a transform closure throws during matching, the matching will abort and the error will be propagated directly to the top-level matching API that's being called, e.g. `Regex.wholeMatch(in:)` and `Regex.prefixMatch(in:)`.
+
+An alternative version of capture is called `TryCapture`, which works in cases where you want to transform the capture, but the transformation may return nil. When a nil is returned, the regex engine backtracks and tries an alternative. `TryCapture` is most useful when you transform a captured substring by calling a failable initializer.
+
+```swift
+enum TransactionKind: String {
+  case credit = "CREDIT"
+  case debit = "DEBIT"
+}
+
+TryCapture {
+  ChoiceOf {
+    "CREDIT"
+    "DEBIT"
+  }
+} transform: {
+  // This initializer may return nil which is why we used TryCapture.
+  TransactionKind(rawValue: String($0))
+}
+```
+
+Reference is a feature that can be used to achieve named captures and named backreferences from textual regexes. Simply state what type the reference will hold on to and you can use it later once you've matched a string to get back a specific capture. Note the type you pass to reference will be whatever the result of a capture's transform is. A capture with no transform always has a reference type of `Substring`.
+
+```swift
+let kind = Reference(Substring.self)
+
+let regex = Capture(as: kind) {
+  ChoiceOf {
+    "CREDIT"
+    "DEBIT"
+  }
+}
+
+let input = "CREDIT"
+if let result = regex.firstMatch(in: input) {
+  print(result[kind]) // CREDIT
+}
+```
+
+When capturing some regex with a reference specified, the reference will refer to the most recently captured content. The reference itself can also be used as a regex to match the most recently captured content, or as a name to look up the result of matching.
+
+```swift
+let a = Reference(Substring.self)
+let b = Reference(Substring.self)
+let regex = Regex {
+  Capture("abc", as: a)
+  Capture("def", as: b)
+  a
+  Capture(b)
+}
+
+if let result = regex.firstMatch(in: input) {
+  print(result[a]) // => "abc"
+  print(result[b]) // => "def"
+}
+```
+
+A regex is considered invalid when it contains a use of reference without it ever being captured in the regex. When this occurs in the regex builder DSL, a runtime error will be reported. Similarly, the use of a reference in a `Regex.Match.subscript(_:)` must have been captured in the regex that produced the match.
+
+<details>
+<summary>API definition</summary>
+
+```swift
+public struct Capture<Output>: RegexComponent {
+  public var regex: Regex<Output> { get }
+}
+
+public struct TryCapture<Output>: RegexComponent {
+  public var regex: Regex<Output> { get }
+}
+```
+
+Below are `Capture` and `TryCapture` initializer variants on capture arity 0. Higher capture arities are omitted for simplicity.
+  
```

Comment:
This would be where I'd start the disclosure triangle.

Reply:
This sentence is already in a collapsible.  The type declarations are both outside the collapsible at the top of the "Capture and reference" section (abbreviated), and within the collapsible (full).

Reply:
The point was that the triangle would happen at this point, not earlier